pkg/kubelet/network/cni/cni.go

```diff
@@ -0,0 +1,232 @@
+	// Perhaps pick the network by some annotation/label on pod, or the namespace.
+	network, err := plugin.selectNetwork(pod)
+	if err != nil {
+		return err
+	}
+	res, err := network.addNetwork(name, namespace, string(id), netns)
+
+	var ip string
+	if res.IP4 != nil {
+		ip = res.IP4.IP.String()
+	} else {
+		ip = res.IP6.IP.String()
+	}
+	// TODO-PAT: check that PodIP can be an IPv6.
+	pod.Status.PodIP = ip
+	return err
+}
+
+func (plugin *cniNetworkPlugin) TearDownPod(namespace string, name string, id kubeletTypes.DockerID) error {
+	netns, err := plugin.host.GetRuntime().GetNetNs(string(id))
+	if err != nil {
+		return err
+	}
+	pod, ok := plugin.host.GetPodByName(namespace, name)
+	if !ok {
+		return fmt.Errorf("pod %q namespace %q: unable to find pod", name, namespace)
+	}
+
+	network, err := plugin.selectNetwork(pod)
+	if err != nil {
+		return err
+	}
+	return network.deleteNetwork(name, namespace, string(id), netns)
+}
+
+func (plugin *cniNetworkPlugin) Status(namespace string, name string, id kubeletTypes.DockerID) (*network.PodNetworkStatus, error) {
+	pod, ok := plugin.host.GetPodByName(namespace, name)
+	if !ok {
+		return nil, fmt.Errorf("pod %q namespace %q: unable to find pod", name, namespace)
+	}
+
+	return &network.PodNetworkStatus{IP: net.ParseIP(pod.Status.PodIP)}, nil
+}
+
+func (network *cniNetwork) addNetwork(podName string, podNamespace string, podInfraContainerID string, podNetnsPath string) (*cniTypes.Result, error) {
+	rt, err := buildRtConfig(podName, podNamespace, podInfraContainerID, podNetnsPath)
+	if err != nil {
+		glog.Errorf("Error adding network: %v", err)
+		return nil, err
+	}
+
+	netconf, cninet := network.NetworkConfig, network.CNIConfig
+	glog.V(2).Infof("About to run with conf.Network.Type=%v, c.Path=%v", netconf.Network.Type, cninet.Path)
+	res, err := cninet.AddNetwork(netconf, rt)
+	if err != nil {
+		glog.Errorf("Error adding network: %v", err)
+		return nil, err
+	}
+
+	return res, nil
+}
+
+func (network *cniNetwork) deleteNetwork(podName string, podNamespace string, podInfraContainerID string, podNetnsPath string) error {
+	rt, err := buildRtConfig(podName, podNamespace, podInfraContainerID, podNetnsPath)
+	if err != nil {
+		glog.Errorf("Error deleting network: %v", err)
+		return err
+	}
+
+	netconf, cninet := network.NetworkConfig, network.CNIConfig
+	glog.V(2).Infof("About to run with conf.Network.Type=%v, c.Path=%v", netconf.Network.Type, cninet.Path)
+	err = cninet.DelNetwork(netconf, rt)
+	if err != nil {
+		glog.Errorf("Error deleting network: %v", err)
+		return err
+	}
+	return nil
+}
+
+func buildRtConfig(podName string, podNs string, podInfraContainerID string, podNetnsPath string) (*libcni.RuntimeConf, error) {
+	glog.V(2).Infof("Got netns path %v", podNetnsPath)
+	glog.V(2).Infof("Using netns path %v", podNs)
+
+	rt := &libcni.RuntimeConf{
+		ContainerID: "cni",
+		NetNS:       podNetnsPath,
+		IfName:      "eth0",
+		Args: [][2]string{
+			{"K8S_POD_NAMESPACE", podNs},
```

Comment:
This is a point of contention.  If we provide this to all plugins, we're encouraging coupling.  But we need to provide SOMETHING and I don't have a better answer.